tools/istio-iptables/pkg/capture/run_test.go

```diff
@@ -337,6 +351,159 @@ func TestSeparateV4V6(t *testing.T) {
 	}
 }
 
+func TestIdempotentEquivalentRerun(t *testing.T) {
+	commonCases := getCommonTestCases()
+	ext := &dep.RealDependencies{
+		CNIMode:          false,
+		NetworkNamespace: "",
+	}
+	iptVer, err := ext.DetectIptablesVersion(false)
+	if err != nil {
+		t.Fatalf("Can't detect iptables version")
+	}
+
+	ipt6Ver, err := ext.DetectIptablesVersion(true)
+	if err != nil {
+		t.Fatalf("Can't detect ip6tables version")
+	}
+
+	for _, tt := range commonCases {
+		t.Run(tt.name, func(t *testing.T) {
+			cfg := constructTestConfig()
+			tt.config(cfg)
+			// Override UID and GID otherwise test will fail in the linux namespace from unshare-go lib
+			cfg.ProxyUID = "0"
+			cfg.ProxyGID = "0"
+			if cfg.OwnerGroupsExclude != "" {
+				cfg.OwnerGroupsInclude = "0"
+			}
+			if cfg.OwnerGroupsInclude != "" {
+				cfg.OwnerGroupsInclude = "0"
+			}
+
+			defer func() {
+				// Final Cleanup
+				cfg.CleanupOnly = true
+				cfg.NoReconcile = false
+				iptConfigurator := NewIptablesConfigurator(cfg, ext)
+				assert.NoError(t, iptConfigurator.Run())
+				residueExists, deltaExists := iptConfigurator.VerifyIptablesState(&iptVer, &ipt6Ver)
+				assert.Equal(t, residueExists, false)
+				assert.Equal(t, deltaExists, true)
+			}()
+
+			// First Pass
+			cfg.NoReconcile = true
+			iptConfigurator := NewIptablesConfigurator(cfg, ext)
+			assert.NoError(t, iptConfigurator.Run())
+			residueExists, deltaExists := iptConfigurator.VerifyIptablesState(&iptVer, &ipt6Ver)
+			assert.Equal(t, residueExists, true)
+			assert.Equal(t, deltaExists, false)
+
+			// Second Pass
+			iptConfigurator = NewIptablesConfigurator(cfg, ext)
+			assert.NoError(t, iptConfigurator.Run())
+		})
+	}
+}
+
+func TestIdempotentUnequaledRerun(t *testing.T) {
+	commonCases := getCommonTestCases()
+	ext := &dep.RealDependencies{
+		CNIMode:          false,
+		NetworkNamespace: "",
+	}
+	iptVer, err := ext.DetectIptablesVersion(false)
+	if err != nil {
+		t.Fatalf("Can't detect iptables version")
+	}
+
+	ipt6Ver, err := ext.DetectIptablesVersion(true)
+	if err != nil {
+		t.Fatalf("Can't detect ip6tables version")
+	}
+
+	for _, tt := range commonCases {
+		t.Run(tt.name, func(t *testing.T) {
+			cfg := constructTestConfig()
+			tt.config(cfg)
+			// Override UID and GID otherwise test will fail in the linux namespace from unshare-go lib
+			cfg.ProxyUID = "0"
+			cfg.ProxyGID = "0"
+			var stdout, stderr bytes.Buffer
+			if cfg.OwnerGroupsExclude != "" {
+				cfg.OwnerGroupsInclude = "0"
+			}
+			if cfg.OwnerGroupsInclude != "" {
+				cfg.OwnerGroupsInclude = "0"
+			}
+
+			defer func() {
+				// Final Cleanup
```

Comment:
is this supposed to bundle all of these to all run AFTER all the commonCases iterations?

Reply:
No, the cleanup is performed after each iteration has finished.
The defer is inside an anonymous function representing a single subtest and the cleanup will be performed at the end of the subtest so that the next subtest can start from with a clean state.